### PR TITLE
Add a test for nss modules and some related fixes

### PR DESCRIPTION
**Pull request by keszybz**
I think a test like this would be useful. Right now a framework to call functions exposed by the nss modules is added. The results are printed, without any real testing. Actual tests for the functionality can be added later on.

The whole nss API is undocumented, so it's hard to say what is right behaviour. Resolving the same thing through different modules and looking at the results can be helpful in divining the expected behaviour.

### CODE REVIEW

File: src/nss-myhostname/nss-myhostname.c [hunk 1]
```diff
@@ -180,8 +180,6 @@ enum nss_status _nss_myhostname_gethostbyname4_r(
         /* Explicitly reset all error variables */
         *errnop = 0;
         *h_errnop = NETDB_SUCCESS;
-        h_errno = 0;
-
```
Comment (poettering) [inline src/nss-myhostname/nss-myhostname.c:183]: Well, there's a certain conflict here between theory and reality...

See e70df46b9721a3d025e7a0b4ffb5893cbde5e55d

**Comment by poettering** [comment]
The h_errno thing should not get merged until glibc is fixed… looks ok otherwise.

File: src/nss-myhostname/nss-myhostname.c (hunk 1 above)
Comment (keszybz) [inline src/nss-myhostname/nss-myhostname.c:183]: Aw, fuck. I'll drop that part.

**Comment by keszybz** [comment]
Any thoughts if we should add some actual tests here? Or is just printing the stuff out good enough for now?

**Comment by poettering** [comment]
Just printing it sounds good enough for now. Real tests are always difficult if network connectivity is required to do sane testing...

For now I'd really just merge this (without the h_errno think) and then leave this for a later commit to be beefed up...

**Comment by poettering** [comment]
btw, please always also drop a comment when you push new code, because otherwise i don't get any notifications for the PR...

**Comment by martinpitt** [comment]
Obviously this worked in CI, but in my local build schroot and also my main desktop this fails with

```
======== resolve ========
_nss_resolve_gethostbyname4_r("localhost") → status=NSS_STATUS_SUCCESS
                   pat=buffer+0x10 errno=999/--- h_errno=1/Unknown host ttl=3600
        "localhost" AF_INET 127.0.0.1 %0
Assertion 'n == 2' failed at src/test/test-nss.c:189, function test_gethostbyname4_r(). Aborting.
Aborted (core dumped)
```

This also seems to be relatively independent of `/etc/hosts`, I was removing lines one by one and I always get the same failure.

According to the `log_info()` format and `NSS_STATUS_SUCCESS` we are in the then-branch in `test_gethostbyname4_r()`. With adding an extra log message I see that `n = print_gaih_addrtuples(pat)` returns 1, thus failing the `n == 2` assertion.

What is this trying to test?

**Comment by keszybz** [comment]
First of all, I screwed up, and forgot to move test-nss to manual_tests. It'd require more work before it was ready for being run automatically.

The test was supposed to test that we get both ::1 and 127.0.0.1. It seems that you're only getting IPv4 address.

**Comment by martinpitt** [comment]
Yeah, my `/etc/hosts` (pretty much as it comes out of the box) has

```
grep localhost /etc/hosts
127.0.0.1   localhost.localdomain   localhost
::1 donald  localhost6.localdomain6 localhost6
::1     ip6-localhost ip6-loopback
```

But I already tried to append `localhost` to the `::1` lines yesterday, and the test still fails.

**Comment by keszybz** [comment]
nss-myhostname and nss-resolve should work without `/etc/hosts` at all. I don't think that file matter here at all.

**Comment by martinpitt** [comment]
Ah! I didn't actually have libnss-myhostname and libnss-resolve installed, but I did that now and it still fails like above.

**Comment by keszybz** [comment]
test-nss uses libnss-\* from the build directory, doesn't touch the system-wide copy. It only uses libnss-dns from the system.

**Pull request merged**